sympy/solvers/tests/test_solveset.py

```diff
@@ -1898,19 +1898,43 @@ def test_solve_expo():
 
 
 # logarithmic tests
-@XFAIL
 def test_logarithmic():
     assert solveset_real(log(x - 3) + log(x + 3), x) == FiniteSet(
-        sqrt(10))
+        -sqrt(10), sqrt(10))
```

Comment:
Is this like the removable singularity case? *IF* this is rewritten as `log((x - 3)*(x + 3))` then see `-sqrt(10)` is a solution. But if it's not rewritten like this then substituting in the `-sqrt(10)` gives an `I*pi` from each term and these don't cancel.

Reply:
>Is this like the removable singularity case?

Yes, it is.
In my opinion `-sqrt(10)` should not be a solution as substituting will cause `log(x - 3)` and `log(x + 3)` to have negative arguments.
As mentioned [here](https://groups.google.com/forum/#!topic/sympy/QOXm9RZdp3c) I am not sure as of now what definition should be considered to remove spurious solutions.

- Should we substitute every solution in each of the log argument and check if it makes it non-real (as it is mentioned in the third example [here](http://www.purplemath.com/modules/solvelog2.htm)), or

- Substitute directly each solution to the original function and see if we get `0` (this can be done using `checksol`). But using `checksol` would sometime cancel out `I*pi` terms which may give wrong solution, like in `log(3*x) - log(1  - x) - log(4*x + 1)`

Reply:
You can return  `ConditionSet(x, Or(x+3 >= 0, x - 3 >=0), {sqrt(10), -sqrt(10)}) -> sqrt(10)`, i.e. keep track of conditions for solution. Again, *I think* that all but one arg must be positive in order for the combined logarithm to be valid. That would make a more complicated condition but at least it is expressible. For 3 args, `Or(And(arg1>=0, arg2>=0), And(arg3>=0, arg2>=0), And(arg1>=0, arg3>=0))`. This could be constructed with `subsets`.

Reply:
Yeah I guess using `ConditionSet` seems to be handy. I will have a look at it.

Reply:
Can we simply use something like: `ConditionSet(x, And(arg1>=0, arg2>=0, arg3>=0), {solutions})` ? where arg1, arg2, arg3 ... will be arguments of every log term in the equation
This works for the following:

```python

# log(x - 3) + log(x + 3)
>>> ConditionSet(x, And(x + 3 >= 0, x - 3 >=0), {sqrt(10), -sqrt(10)})
{sqrt(10)}

# log(3*x) - log(1 - x) - log(4*x + 1)
>>> ConditionSet(x, And(3*x>=0, 1 - x>=0, 4*x + 1>=0), {S(1)/2, -S(1)/2})
{1/2}
```

Reply:
What about `log(1-x) - log(3*x) - log(4*x + 1) = 0` for which `x = -1/2` is a solution which makes only two args 0?

Reply:
> What about log(1-x) - log(3*x) - log(4*x + 1) = 0 for which x = -1/2 is a solution which makes only two args 0?

[Wolfram alpha](http://www.wolframalpha.com/input/?i=solve(log(1-x)+-+log(3*x)+-+log(4*x+%2B+1),+x)) gives `1/6` as the only solution to this equation. It rules out `-1/2` maybe because `-1/2` makes `log(3*x)` and `log(4*x + 1)` negative.

To see we are on the same page what should be followed while removing bad solutions.

- Any of the solution makes any of the log arg negative should not be included, or

- Substituting each solution to the original equation and see it evaluates to `0`.

I think wolfram alpha follows the first criteria, as evident from these two equations:
[log(3*x) - log(1 - x) - log(4*x + 1)](http://www.wolframalpha.com/input/?i=solve(log(3*x)+-+log(1+-+x)+-+log(4*x+%2B+1),+x)) and [log(1-x) - log(3*x) - log(4*x + 1)](http://www.wolframalpha.com/input/?i=solve(log(1-x)+-+log(3*x)+-+log(4*x+%2B+1),+x))

The `ConditionSet` case I gave above is valid considering the first case. What do you suggest?

```python

>>> ConditionSet(x, And(1 - x >0, 3*x > 0, 4*x + 1), {-1/2, 1/6})
1/6
```

Reply:
Although `log(a) + log(b) = log(a*b)` when either a or b is positive, `log(a) - log(b)` is only log(a/b)` when both are positive. And `log(a) + log(b) + log(c)` will only be `log(a*b*c)` when all three arguments are positive or else there would be an extra multiple of `I*pi` in the former case.

A solution to `log(a) + log(b) = 0` will only be found when both are positive. If one or both are negative there will be a non-cancelling imaginary part. The same applies to the case of 3 terms: if ever any of them is negative there will be a non-cancelling `I*pi` in the numerator or denominator that will make the solution not be zero. 

So, lacking a counterexample, it would seem that your simpler criteria that all args be positive is valid.

Reply:
Talking about removing unwanted solutions @jksuom suggested that we can allow solutions which we get by rewriting an equation in its equivalent form. Taking an example: `sin(x)/x = 1`,  `x` can have `0` as the solution because we can write the original equation to `sin(x) = x`

Reply:
I thought this was a case of a removable singularity. If division by 0 is not defined then `sin(x)/x` is undefined even though the limit as `x -> 0` is defined.